openedx/features/discounts/applicability.py

```diff
@@ -55,6 +57,15 @@ def can_receive_discount(user, course):  # pylint: disable=unused-argument
     if DiscountRestrictionConfig.disabled_for_course_stacked_config(course):
         return False
 
+    # Don't allow users who have enrolled in any courses in non-upsellable
+    # modes
+    if CourseEnrollment.objects.filter(user=user).exclude(mode__in=CourseMode.UPSELL_TO_VERIFIED_MODES).exists():
```

Comment:
Do we want to unify this logic with our definition of "has not purchased" aka _has_non_audit_enrollments_ in https://github.com/edx/edx-platform/blob/master/lms/djangoapps/experiments/utils.py?

Reply:
Yeah. I feel better about using this logic, because it covers things like users w/ old `honor` courses, for instance. But it makes sense to update that other spot, too.